### ISSUE TITLE
Feature request: Caret line position vs. paragraph position

### ISSUE THREAD

**Issue by DaveJarvis** [TomasMikula/RichTextFX#386]
The methods [lineStart](http://fxmisc.github.io/richtext/javadoc/0.7-M2/org/fxmisc/richtext/model/NavigationActions.html#lineStart-org.fxmisc.richtext.model.NavigationActions.SelectionPolicy-) and [lineEnd](http://fxmisc.github.io/richtext/javadoc/0.7-M2/org/fxmisc/richtext/model/NavigationActions.html#lineEnd-org.fxmisc.richtext.model.NavigationActions.SelectionPolicy-) move the cursor to the start and end of the _paragraph_, not the line. Their names should be changed to `paragraphStart` and `paragraphEnd` to avoid confusion, as well as renaming [selectLine](http://fxmisc.github.io/richtext/javadoc/0.7-M2/org/fxmisc/richtext/model/NavigationActions.html#selectLine--) to `selectParagraph`.

When pressing either `Home` or `End`, the caret position should change relative to the current line, not the current paragraph, which is what users have come to expect (e.g., the text area used to write this message has typical home/end key behaviour). There are a few related issues: #384 and #360.

It would be convenient if there was a `getLineNumber()` method that indicated the line number within the current paragraph, such as:

```
StyleClassedTextArea editor = // ...
Paragraph p = editor.getCurrentParagraph();
int n = p.getLineNumber();
```

Or:

```
StyleClassedTextArea editor = // ...
int n = editor.getCurrentLineNumber(); // delegates to paragraph
```

Once `lineStart` and `lineEnd` provide line-level values (instead of paragraph), it implies that writing a key interceptor for cutting a line (e.g., typing `Ctrl+x` as per IDE behaviour) should be as simple as:

```
editor.lineStart(); // should clear by default
int start = editor.getCaretPosition();
editor.lineEnd(); // should clear by default
int end = textArea.getCaretPosition();
editor.deleteText( start, end );
```

Or, pie-in-the-sky, to avoid extraneous UI updates:

```
int start = editor.getCaretLineStartColumn(); 
int end = editor.getCaretLineEndColumn();
editor.deleteText( start, end );
```

This would also make it trivial to override the default Home/End behaviour to move the cursor to the start and end of lines, rather than paragraphs.

**Comment by JordanMartinez** [comment]
> The methods lineStart and lineEnd move the cursor to the start and end of the paragraph, not the line. Their names should be changed to paragraphStart and paragraphEnd to avoid confusion, as well as renaming selectLine to selectParagraph.

Good point. 

> When pressing either Home or End, the caret position should change relative to the current line, not the current paragraph, which is what users have come to expect (e.g., the text area used to write this message has typical home/end key behaviour). There are a few related issues: #384 and #360.

In other words, a paragraph that spans multiple lines is wrapped should move the caret to the start of the current line, not the paragraph itself. However, does a normal JavaFX `TextArea` do that?

> It would be convenient if there was a getLineNumber() method that indicated the line number within the current paragraph

Such a request would probably need to expose more of `TextFlow`s API in `TextFlowExt`. Also a good point.

**Comment by DaveJarvis** [comment]
> In other words, a paragraph that spans multiple lines is wrapped should move the caret to the start of the current line, not the paragraph itself. However, does a normal JavaFX TextArea do that?

Correct; however, I'm not how JavaFX TextArea behaves -- I imagine that it distinguishes lines from paragraphs as that's been typical UI behaviour for editing text since [1978](https://en.wikipedia.org/wiki/WordStar) or so. ;-)

**Comment by JordanMartinez** [comment]
> I imagine that it distinguishes lines from paragraphs as that has been typical UI behaviour for editing text since 1978 or so. ;-)

:smile: 

I just tested it and you're right.

**Comment by DaveJarvis** [comment]
> StyleClassedTextArea editor = // ...
> int n = editor.getCurrentLineNumber(); // delegates to paragraph

Probably don't want to delegation here. Asking the editor's line number should give the line number from the top of the text (starting from line 1), not the current paragraph. Also, using natural numbers seems like a good fit here, as opposed to whole numbers, for line numbers.

**Comment by JordanMartinez** [comment]
This was closed by #419. See additional comments in #417.
